bin/convert2vulyk.py

```diff
@@ -67,43 +97,156 @@ def parse_bsf(bsf_data: str) -> list:
+    result: list = []
     for m in ln_ptrn.finditer(data):
         bsf = BsfInfo(m.group(1), m.group(2), int(m.group(3)), int(m.group(4)), m.group(5).strip())
         result.append(bsf)
     return result
 
 
-def convert_bsf_2_vulyk(text: str, bsf_markup: str) -> dict:
+def simple_tokenizer(text: str) -> list[list[str]]:
+    """
+    Given whitespace/newline tokenized text, return the
+    list of sentences (where each sentence is made of tokens)
+    using whitespaces and newlines
+    """
+
+    doc: list[list[str]] = []
+    if text:
+        for s in text.split("\n"):
+            doc.append(s.split(" "))
+
+    return doc
+
+
+def read_and_tokenize(text: str, fmt: str, tokenizer: TokenizationType) -> list[list[str]]:
+    if fmt == "json":
+        assert tokenizer == TokenizationType.NOOP, "Json is meant to be already tokenized"
+        return json.loads(text)
+    else:
+        assert tokenizer != TokenizationType.NOOP, "You cannot keep texts not-tokenized"
+        if tokenizer == TokenizationType.WHITESPACE:
+            return simple_tokenizer(text)
+        elif tokenizer == TokenizationType.TOKENIZE_UK:
+            return list(chain(*tokenize_text(text)))
+
+    return []  # Calm down, mypy
+
+
+def reconstruct_tokenized(tokenized_text: list[list[str]]) -> Generator[AlignedToken, None, None]:
+    """
+    Accepts tokenized text [["sent1_word1", "sent1_word2"], ["sent2_word2"]]
+    and normalizing the spaces in the text according to the punctuation.
+    Returns an iterator over AlignedToken, where each token has the information
+    on the original position and updated position
+    """
+    SPACES_BEFORE: str = "([“«"
+    NO_SPACE_BEFORE: str = ".,:!?)]”»"
+
+    orig_pos: int = 0
+    adj_pos: int = 0
+
+    for s_idx, s in enumerate(tokenized_text):
+        if s_idx > 0:
+            yield AlignedToken("\n", (orig_pos, orig_pos + 1), (adj_pos, adj_pos + 1))
+            orig_pos += 1
+            adj_pos += 1
+
+        prev_token: str = ""
+        for w_idx, w in enumerate(s):
+            w_stripped = w.strip()
+
+            if not w_stripped:
+                # If original text contained a space(-es), let's adjust original position for it
+                # + one space after
+                orig_pos += len(w)
+                if w_idx > 0:
+                    orig_pos += 1
+
+                continue
+
+            if w_idx > 0:
+                if w_stripped not in NO_SPACE_BEFORE and not prev_token in SPACES_BEFORE:
+                    yield AlignedToken(" ", (orig_pos, orig_pos + 1), (adj_pos, adj_pos + 1))
+                    orig_pos += 1
+                    adj_pos += 1
+                else:
+                    # If we are omitting the space (for example, before comma), we
+                    # adjusting original position as if it's there
+                    orig_pos += 1
+
+            yield AlignedToken(w_stripped, (orig_pos, orig_pos + len(w)), (adj_pos, adj_pos + len(w_stripped)))
```

Comment:
What's your reasoning behind using Generator here? Have you noticed some performance issues?

Considering that the whole thing is in memory already, I'd say that we should not notice any major improvements with using generators here. If we were loading file by file - that would be a different story.

I don't see any point to reverse the decision right now but just curious about your reasons.

Reply:
Well, there are some:

- For me personally it's easier to just yield the token by token, rather than accumulating it somewhere
- I've been playing with idea of supplying the input tokens one by one to perform alignment outside of the function. Later I gave up.
- Some tasks (outside of the NER) requires massive processing of big texts, so maybe I'll reuse it somewhere else someday